File: 100. Same Tree/sol.go
package sameTree



type TreeNode struct {
	Val int
	Left *TreeNode  
	Right *TreeNode 
}


// dfs
func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil {
        return true
    }
    if p == nil || q == nil {
        return false
    }
    if p.Val != q.Val {
        return false
    }
    left := isSameTree(p.Left, q.Left)
    right := isSameTree(p.Right, q.Right)
    return left && right 
}

// bfs
func isSameTreeBFS(p *TreeNode, q *TreeNode) bool {

    stack := [][]*TreeNode {[]*TreeNode {p, q}}
    for len(stack) > 0 {
        index := len(stack) - 1
        a, b := stack[index][0], stack[index][1]
        stack = stack[:index]
        if a != nil || b != nil {
            if a == nil || b == nil {
                return false
            }
            if a.Val != b.Val {
                return false
            }
            stack = append(stack, []*TreeNode {a.Left, b.Left})
            stack = append(stack, []*TreeNode {a.Right, b.Right})
        }
    }


    return true
}

File: 100. Same Tree/sol.py
from typing import Optional

class TreeNode:
    def __init__(self, val: int = 0, left: 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val 
        self.left = left 
        self.right = right 

class Solution: # dfs
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        def dfs(t1:Optional[TreeNode], t2: Optional[TreeNode]) -> bool:
            if not t1 and not t2:
                return True
            if not t1 or not t2:
                return False
            if t1.val != t2.val:
                return False
            left = dfs(t1.left, t2.left)
            right = dfs(t1.right, t2.right)
            return left and right 
        return dfs(p, q)



class Solution2: #bfs
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        stack = [[p, q]]
        while len(stack) > 0 :
            a, b = stack.pop()
            if a or b:
                if not a or not b:
                    return False
                if a.val != b.val:
                    return False
                stack.append([a.left, b.left])
                stack.append([a.right, b.right])
        return True
        